authz/src/main/java/io/grpc/authz/AuthorizationPolicyTranslator.java

```diff
@@ -0,0 +1,197 @@
+      }
+      if (pathsSet.getRulesCount() > 0) {
+        andSet.addRules(Permission.newBuilder().setOrRules(pathsSet.build()).build());
+      }
+    }
+    List<Map<String, ?>> headersList = JsonUtil.getListOfObjects(request, "headers"); 
+    if (headersList != null) {
+      Permission.Set.Builder headersSet = Permission.Set.newBuilder();
+      for (Map<String, ?> header: headersList) {           
+        headersSet.addRules(parseHeader(header));
+      }
+      if (headersSet.getRulesCount() > 0) {
+        andSet.addRules(Permission.newBuilder().setAndRules(headersSet.build()).build());
+      }
+    }
+    if (andSet.getRulesCount() == 0) {
+      return Permission.newBuilder().setAny(true).build();
+    }
+    return Permission.newBuilder().setAndRules(andSet.build()).build();
+  }
+
+  private static Map<String, Policy> parseRules(
+      List<Map<String, ?>> objects, String name) throws IllegalArgumentException {
+    Map<String, Policy> policies = new LinkedHashMap<String, Policy>();
+    for (Map<String, ?> object: objects) {
+      String policyName = JsonUtil.getString(object, "name");
+      if (policyName == null) {
+        throw new IllegalArgumentException("rule \"name\" is absent.");
+      }
+      List<Principal> principals = new ArrayList<>();
+      Map<String, ?> source = JsonUtil.getObject(object, "source");
+      if (source != null) {
+        principals.add(parseSource(source));
+      } else {
+        principals.add(Principal.newBuilder().setAny(true).build());
+      }
+      List<Permission> permissions = new ArrayList<>();
+      Map<String, ?> request = JsonUtil.getObject(object, "request");
+      if (request != null) {
+        permissions.add(parseRequest(request));
+      } else {
+        permissions.add(Permission.newBuilder().setAny(true).build());
+      }
+      Policy policy = 
+          Policy.newBuilder()
+          .addAllPermissions(permissions)
+          .addAllPrincipals(principals)
+          .build();
+      policies.put(name + "_" + policyName, policy);
+    }
+    return policies;
+  }
+
+  /** 
+  * Translates a gRPC SDK authorization policy in JSON string to Envoy RBAC policies.
+  * On success, will return one of the following -
+  * 1. One allow RBAC policy or,
+  * 2. Two RBAC policies, deny policy followed by allow policy.
+  * If the policy cannot be parsed or is invalid, an exception will be thrown.
+  */
+  public static List<RBAC> translate(String authorizationPolicy) 
+  throws IllegalArgumentException, IOException {
+    @SuppressWarnings("unchecked")
+    Map<String, ?> json = (Map<String, ?>)JsonParser.parse(authorizationPolicy);
+    String name = JsonUtil.getString(json, "name");
+    if (name == null) {
+      throw new IllegalArgumentException("\"name\" is absent.");
+    }
+    List<RBAC> rbacs = new ArrayList<>();
+    List<Map<String, ?>> objects = JsonUtil.getListOfObjects(json, "deny_rules");
+    if (objects != null) {
```

Comment:
I'm a bit concerned about null vs empty behavior. I agree they look like the'd eventually produce the same RBAC decisions, but I'd feel more comfortable if this had an empty check (or if null had an empty policy list).

Reply:
Sure thing. Added check for empty list.